module/zfs/dnode.c

```diff
@@ -1949,18 +1949,25 @@ static void
 dnode_dirty_l1range(dnode_t *dn, uint64_t start_blkid, uint64_t end_blkid,
     dmu_tx_t *tx)
 {
-	dmu_buf_impl_t db_search;
+	dmu_buf_impl_t *db_search;
 	dmu_buf_impl_t *db;
 	avl_index_t where;
 
+#ifdef ZFS_DEBUG
+	db_search = kmem_alloc(sizeof (dmu_buf_impl_t), KM_SLEEP);
+#else
```

Comment:
This works, but I wonder if we could make this a little cleaner by putting the level/blikd/state (which are used in the AVL comparator) all together and then allocating just that part of the dbuf on the stack?  But maybe that's more tricky than it's worth, since this isn't a performance critical path.

Reply:
I was thinking about that too, but in the end all those function parameters are (most often) just 64 bit numbers few bytes; the other one possible improvement I was thinking about is allocating a buffer for where and db_search together, which would save even more stack space. In the end this was enough, so I left it at that. I can re-do this, whichever way we end up liking in the end, no problem.

Reply:
My feeling is that keeping this simple and always doing a `kmem_alloc()` (debug and non-debug) would be preferable.